d2layouts/d2grid/grid_diagram.go

```diff
@@ -0,0 +1,81 @@
+package d2grid
+
+import (
+	"strconv"
+
+	"oss.terrastruct.com/d2/d2graph"
+)
+
+type gridDiagram struct {
+	root    *d2graph.Object
+	nodes   []*d2graph.Object
+	rows    int
+	columns int
+
+	rowDominant bool
+
+	width  float64
+	height float64
+}
+
+func newGridDiagram(root *d2graph.Object) *gridDiagram {
+	gd := gridDiagram{root: root, nodes: root.ChildrenArray}
+	if root.Attributes.Rows != nil {
+		gd.rows, _ = strconv.Atoi(root.Attributes.Rows.Value)
+	}
+	if root.Attributes.Columns != nil {
+		gd.columns, _ = strconv.Atoi(root.Attributes.Columns.Value)
+	}
+
+	// compute exact row/column count based on values entered
+	if gd.columns == 0 {
+		gd.rowDominant = true
+	} else if gd.rows == 0 {
+		gd.rowDominant = false
+	} else {
+		// if keyword rows is first, rows are primary, columns secondary.
+		if root.Attributes.Rows.MapKey.Range.Before(root.Attributes.Columns.MapKey.Range) {
+			gd.rowDominant = true
+		}
+
+		// rows and columns specified, but we want to continue naturally if user enters more nodes
+		// e.g. 2 rows, 3 columns specified + g node added: │ with 3 columns, 2 rows:
+		// . original  add row   add column                 │ original  add row   add column
+		// . ┌───────┐ ┌───────┐ ┌─────────┐                │ ┌───────┐ ┌───────┐ ┌─────────┐
+		// . │ a b c │ │ a b c │ │ a b c d │                │ │ a c e │ │ a d g │ │ a c e g │
+		// . │ d e f │ │ d e f │ │ e f g   │                │ │ b d f │ │ b e   │ │ b d f   │
+		// . └───────┘ │ g     │ └─────────┘                │ └───────┘ │ c f   │ └─────────┘
+		// .           └───────┘ ▲                          │           └───────┘ ▲
+		// .           ▲         └─existing nodes modified  │           ▲         └─existing nodes preserved
+		// .           └─existing rows preserved            │           └─existing rows modified
+		capacity := gd.rows * gd.columns
+		for capacity < len(gd.nodes) {
```

Comment:
I'm still not sure about this. While I get that this is quite convenient, I think it can be hard to track down. If users want 3 columns and many rows, they should just specify `columns: 3`

Reply:
columns: 3 on its own is different from having both columns: 3 and rows: 3. when you also specify rows, all nodes in a row must have the same height.

![Screen Shot 2023-04-06 at 1 20 15 PM](https://user-images.githubusercontent.com/85081687/230485855-a6e66aaa-0465-4e85-8f72-8f0eac4d9391.png)

Reply:
I see, what if we just had a warning then?
I think it is somewhat important to let users know that there are more nodes than the specified capacity

Reply:
This makes it really simple to add or remove nodes without having to update rows and columns all the time: 

https://user-images.githubusercontent.com/85081687/230491900-6b81f289-15cc-4e55-a4f6-8acb6d38a2fa.mp4

We don't have warnings right now afaik, only errors, but a warning might be ok. don't know if it is needed